app/models/piece.rb

```diff
@@ -79,21 +65,83 @@ def moved?(x, y)
   end
 
   # Returns true if the coordinates provided
-  # have the same x-axis value.
-  def horizontal_move?(x)
-    x_distance(x) == 0
+  # have the same x-axis value and there are no
+  # pieces in between.
+  def clear_horizontal_move?(x, y)
+    return false unless y_distance(y) == 0
+    distance = x_distance(x)
+    path_clear?(x, y, distance)
   end
 
   # Returns true if the coordinates provided
-  # have the same y-axis value.
-  def vertical_move?(y)
-    y_distance(y) == 0
+  # have the same y-axis value and there are no
+  # pieces in between.
+  def clear_vertical_move?(x, y)
+    return false unless x_distance(x) == 0
+    distance = y_distance(y)
+    path_clear?(x, y, distance)
   end
 
   # Returns true if the coordinates provided
   # are the same distance away from the
-  # origin point along both axis.
-  def diagonal_move?(x, y)
-    x_distance(x) == y_distance(y)
+  # origin point along both axis and there are no
+  # pieces in between.
+  def clear_diagonal_move?(x, y)
+    return false unless x_distance(x) == y_distance(y)
+    distance = x_distance(x)
+    path_clear?(x, y, distance)
+  end
+
+  # Class constants used to determine path direction
+  # for obstruction logic.
+  RIGHT = 1
+  LEFT = -1
+  DOWN = 1
+  UP = -1
+
+  # Returns a hash with 1 or -1 values for
+  # x and y based on whether those values are
+  # increasing or decreasing towards destination
+  # to determine direction of path along both axis.
+  def path_direction(x, y)
+    direction = {}
+    direction[:x] = if x_position < x then RIGHT
+                    elsif x_position > x then LEFT
+                    else 0
+                    end
+    direction[:y] = if y_position < y then DOWN
+                    elsif y_position > y then UP
+                    else 0
+                    end
+    direction
+  end
+
+  # Returns an array of x-y coordinate subarrays
+  # between origin and destination.
+  def generate_path_coordinates(x, y, distance)
+    coordinate_sets = []
+    direction = path_direction(x, y)
+    (distance - 1).times do |i|
+      i += 1
+      coordinate_sets << [
+        x_position + i * direction[:x],
+        y_position + i * direction[:y]
```

Comment:
Even though here it doesn't truly matter due to just -1 or 1, had our numbers been slightly different you would have had some shitty results. Take a look at Ruby operator precedence (aka order of operations) - http://www.techotopia.com/index.php/Ruby_Operator_Precedence